File: C++/Baekjoon/스도쿠.cpp
// https://www.acmicpc.net/problem/2580
// 2580.스도쿠 : 2020.10.11

#include <iostream>
#include <vector>
#define pii pair<int,int>
#define vi vector<int>
#define SIZE 9

using namespace std;

// 스도쿠 판
vector<vi> sudoku(SIZE, vi(SIZE));
// 아직 채워지지 않은 좌표
vector<pii> point;

// depth 번째 채워지지 않은 좌표를 채운다
bool fillSudoku(int depth){
    // 모든 비어있는 좌표를 채운 경우 출력
    if(depth == point.size()){
        for(int i=0; i<SIZE; i++){
            for(int j=0; j<SIZE; j++){
                cout<<sudoku[i][j]<<' ';
            }cout<<'\n';
        }
        return true;
    }

    // 현재 검사하는 곳의 좌표
    int dy = point[depth].first, dx = point[depth].second;
    // i번째 숫자를 사용하고 있는지 저장
    vector<bool> isUsedNum(SIZE + 1);

    // 스도쿠 가로, 세로 검사
    for(int i=0; i<SIZE; i++){
        isUsedNum[sudoku[dy][i]] = true;
        isUsedNum[sudoku[i][dx]] = true;
    }

    int y = (dy / 3) * 3;
    int x = (dx / 3) * 3;

    // 3 x 3 칸 검사
    for(int i=0; i<3; i++){
        for(int j=0; j<3; j++){
            isUsedNum[sudoku[y+i][j+x]] = true;
        }
    }

    // 넣을 수 있는 수를 가지고 검사
    for(int i=1; i<=SIZE; i++){
        if(!isUsedNum[i]){
            sudoku[dy][dx] = i;
            if(fillSudoku(depth + 1)) return true;
            sudoku[dy][dx] = 0;
        }
    }
    return false;
}

int main(){
    ios_base::sync_with_stdio(false); cin.tie(NULL);

    for(int i=0; i<SIZE; i++){
        for(int j=0; j<SIZE; j++){
            cin>>sudoku[i][j];
            if(!sudoku[i][j]){
                point.push_back({i, j});
            }
        }
    }

    fillSudoku(0);
}

File: python/baekjoon/스도쿠.py
# https://www.acmicpc.net/problem/2580
# 2580.스도쿠 : 2020.10.11

import sys

SIZE = 9

# 입력을 받아 리스트에 정수로 저장
sudoku = [list(map(int, sys.stdin.readline().split())) for i in range(SIZE)]
point = []

# 비어있는 곳의 좌표를 저장
for i in range(SIZE):
    for j in range(SIZE):
        if not sudoku[i][j]:
            point.append((i, j))

def fillSudoku(depth):
    # 스도쿠가 풀린 경우 출력
    if depth == len(point):
        for i in range(SIZE):
            print(' '.join(map(str, sudoku[i])))
        return True

    # 현재 자리에 넣을 수 있는 숫자를 찾기 위한 리스트
    isUsedNum = [False for i in range(SIZE + 1)]

    dy = point[depth][0]
    dx = point[depth][1]

    # 가로, 세로 검사
    for i in range(SIZE):
        isUsedNum[sudoku[dy][i]] = True
        isUsedNum[sudoku[i][dx]] = True

    y = (dy // 3) * 3;
    x = (dx // 3) * 3;

    # 3 x 3 검사
    for i in range(3):
        for j in range(3):
            isUsedNum[sudoku[y+i][x+j]] = True

    # BackTracking 방법으로 다음번 비어있는 곳 채우기
    for i in range(1, SIZE + 1):
        if not isUsedNum[i]:
            sudoku[dy][dx] = i
            if fillSudoku(depth + 1):
                return True
            sudoku[dy][dx] = 0

    return False

fillSudoku(0)